community,wg,code-quality: establish working group code-quality

dhiller opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Create wg charter and add sigs.yaml entry for wg-code-quality.

Update sig-list.md to reflect changes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #327 

**Special notes for your reviewer**:

This PR needs to get rebased after https://github.com/kubevirt/community/pull/346 got in

dankenigsberg commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,52 @@
+  * This may include presentations, documentation, lectures, certifications, linters, etc.
+* Enforcement of selected standards.
+  * Through automation and active intervention in code reviews.
+  * Through codebase infrastructure.
```
```suggestion
  * Through codebase infrastructure. E.g reusable libraries.
```
(the existing text was not clear to me without the e.g)

dankenigsberg commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,52 @@
+* Management:
+  * Prioritizing tasks.
+  * Focusing on selected topics.
+* Monitor and identification of code-quality issues.
```
```suggestion
* Monitoring and identification of code-quality issues.
```

dankenigsberg reviewed (changes requested):
(partial review)

EdDev commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,52 @@
+
+## Deliverables
+
+The inter-SIG process definition will be manifested in a document that defines the process itself, also how and when it is executed. 
```
I think the deliverables should be the result of the scope points (objectives in the current naming), more or less.

I like the general wording of the device-manager WG:

> The WG will coordinate the delivery of KEPs and their implementations by the participating SIGs. Interim artifacts will include documents capturing use cases, requirements, and designs; however, all of those will eventually result in KEPs and code owned by SIGs.

We could re-word it a bit to fit better to our naming (or lack of it) and scope:

> The WG will coordinate the prioritization and delivery of design-proposals, documentation, implementations and other content that aims to improve the code quality of the project.

EdDev reviewed (commented):
Thank you @dhiller for pushing this forward.
This is basically formalizing an ongoing active effort, allowing us to centralize, prioritize and learn how to make out code base cleaner/better.

EdDev commented:
Regarding the the chair and roles in general assignment, it will be useful to have a ref or and explanation (and example?) of what that means. Then it will be clearer to members what they need to do and estimate the needed investment.

I think we need a minimum of 2 chairs, preferable 3.
I am not really clear on how a nomination is done when a WG is created, it is a bit odd.
In any case, given my high interest in this WG success and my assumptions on what the role requires, please consider me as one of the chairs.

dhiller commented:
> Thank you @dhiller for pushing this forward. This is basically formalizing an ongoing active effort, allowing us to centralize, prioritize and learn how to make out code base cleaner/better.

@EdDev thanks for the review - I've changed wording accordingly, hope that this conveys better what we want to achieve.

Also I've updated with the new links to the meeting room, and added a paragraph for disbanding. This I've left vague intentionally.

One note though: this WG IMHO is about creating the process (which is kind of a meta activity) that will serve as a vehicle for continuous improvement. I think that we can iterate on that process, the docmentation and the automation continuously, so therefore the WG might disband after a certain "good enough" set of processes and documentation is in place.

dhiller commented:
/hold to wait for rebase after https://github.com/kubevirt/community/pull/346 merged

EdDev commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,59 @@
+  * Focusing on selected topics.
+* Monitoring and identification of code-quality issues.
+
+The WG will therefore work on processes through which the above goals can be achieved.
```
> work on processes

I am unsure if this is clear enough to keep it in the text.

As I see it, the WG should work on providing the means to improve the quality until a point where it is well integrated in the projects. The points above hints what we expect to have in the end.

I do not think it is limited to creating processes, it is there to facilitate the effort to create content, refactor code, etc. As it is already mentioned in the points above, I think this sentence is redundant.

dhiller replied:
I have changed the In Scope section by changing the last sentence to what you pointed out before, which IMHO brings it down into an actionable goal. WDYT?

EdDev commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,59 @@
+
+The Code-Quality Working Group has two goals:
+
+* (short-term) **focus on the improvement of code quality** and
+* (long-term) **create an inter-SIG process for identifying, agreeing on and addressing code quality issues**
```
For me, line 14 is summarizing this and describe why the WG is needed.
The long-term goal here, which mentions a processes, is not very clear to me.
I would expect the WG to have artifacts from which the quality will be improved and sustained over time.

This may be partially achieved through processes, but I am unsure if that is the "solution" or how it will look-like.

dhiller replied:
I have changed the In Scope section by changing the last sentence to what you pointed out before, which IMHO brings it down into an actionable goal. WDYT?

EdDev replied:
I guess I do not see this inter-SIG process in my mind and therefore have hard time to understand and agree on it being a goal. I feel this section is redundant as everything is described in the next section.

However, this is a small part compared to all this charter content, so I should not get stuck on it.
We can raise it later and adjust per need.

EdDev reviewed (commented):
> One note though: this WG IMHO is about creating the process (which is kind of a meta activity) that will serve as a vehicle for continuous improvement. I think that we can iterate on that process, the docmentation and the automation continuously, so therefore the WG might disband after a certain "good enough" set of processes and documentation is in place.

The term "process" is a bit to vague to me.
I would expect the WG to facilitate and manage the quality effort until enough tooling, documentation, code structure, reviewer and maintainer awareness is achieved. Once the code quality reached a reasonable state, the WG can be dismissed.

My remaining comments is all about this "process" I guess.
It is not a stopper though, we can and probably should sync on the expectation in the next WG meeting.
Per initial discussions we can keep on updating the charter to reflect what we learn and where we aim to go with this.

dhiller commented:
> > One note though: this WG IMHO is about creating the process (which is kind of a meta activity) that will serve as a vehicle for continuous improvement. I think that we can iterate on that process, the docmentation and the automation continuously, so therefore the WG might disband after a certain "good enough" set of processes and documentation is in place.
> 
> The term "process" is a bit to vague to me. I would expect the WG to facilitate and manage the quality effort until enough tooling, documentation, code structure, reviewer and maintainer awareness is achieved.

That is a good sentence, I will add that to the charter if you don't mind.

> Once the code quality reached a reasonable state, the WG can be dismissed.

I disagree with that the WG should disband after the code quality has reached a reasonable state. I think that it should disband after the above described facilitation and management is complete - meaning that tooling, docs etc. are in effect in order to not let the quality deteriorate below a certain lower boundary.

dhiller commented:
@EdDev I've updated the charter - maybe that is good enough for a start? What do others think?

iholder101 commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,64 @@
+
+## Timelines and Disbanding
+
+The WG will decide at a certain point that the basic goal is achieved - when there is good enough documentation, tooling and metrics about the targets described above available. In that case it can decide to disband itself.
```
I think that this kind of phrasing could be a good compromise / starting point.
From my POV, let's achieve the goals above and then re-asses whether we still need this WG around or not. We might choose to disband it, or change its goal. As an example, lowering the meeting frequency from bi-weeky to monthly could make sense.

I think this shouldn't be a blocker for creating this WG.

dhiller replied:
I agree that a fundamental part of this effort is the continuous improvement and refinement of the goals - this will also define the outcomes at large.

iholder101 commented on sigs.yaml:
```diff
@@ -289,6 +289,33 @@ workinggroups:
           name: "Howard Zhang"
           company: ARM
+    - dir: wg-code-quality
+      name: Code-Quality Working Group
```
Since we've decided to change the github label to `kind/cleanup`, should we align the name of the WG accordingly?

dhiller replied:
Not sure about that - code-quality does bring a certain context where the phrase cleanup can mean anything.

dhiller commented on sigs.yaml:
```diff
@@ -289,6 +289,33 @@ workinggroups:
+        The Code-Quality Working Group aims to create an inter-SIG process for identifying, agreeing on and addressing code quality issues
+      label: wg/code-quality
+      leadership:
+        chairs:
```
The role of a WG chair is defined in the [Community Membership Policy](https://github.com/kubevirt/community/blob/main/membership_policy.md#working-group-wg-chair). What are we missing there?

Or do you mean that the WG chair role for the code-quality needs a better definition?

I agree that the mission statement might need an update now though.

iholder101 replied:
> The role of a WG chair is defined in the [Community Membership Policy](https://github.com/kubevirt/community/blob/main/membership_policy.md?rgh-link-date=2024-11-07T09%3A14%3A52Z#working-group-wg-chair). What are we missing there?

Ah, looks good to me :+1: 
I thought that maybe we should define chair specifically for this WG, but looking at the link you've added I'm not sure that's needed TBH.

Thank you! You're welcome to add me as a chair :)

orelmisan commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,64 @@
+### Special Interest Groups
+
+* All SIGS claiming code ownership and producing code, currently
+  * SIG Compute,
+  * SIG Network and
+  * SIG Storage
+* sig-ci
```
nit:
```suggestion
  * SIG Compute
  * SIG Network
  * SIG Storage
  * SIG-CI
```

orelmisan commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,64 @@
+## Deliverables
+
+The inter-SIG process definition will be manifested in one or more documents that define the quality effort:
+* what tooling and what documentation is required,
+* what metrics for code structure should be established and how it will be implemented,
```
nit:
```suggestion
* what tooling and what documentation is required
* what metrics for code structure should be established and how it will be implemented
```

orelmisan commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,64 @@
+
+## In scope
+
+The main objective of this effort is to continuously improve the quality of the project codebase.
```
Since the project is quite big, is there an idea of where to start?

EdDev replied:
The charter is focused on *what* is needed rather on *how*.
The definition of the scope just provides the understanding of what the WG is considering under its influence/scope.

As @dhiller pointed out, we will discuss how to do things and which areas to invest first in our bi-weekly meetings and through regular communications.

orelmisan reviewed (commented):
Thank you @dhiller for kicking off this effort

lyarwood commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,64 @@
+- @iholder101
+
+#### Liaisons
+- @dhiller
```
This isn't important but it might be useful to have someone from `sig-storage` as a chair or liaison?

EdDev replied:
It will be great to have another chair, 3 sounds ideal.
The WG needs to communicate with other SIGs to apply some of its work items, however, we do not require representatives from SIGs to have a role in the WG. At least I do not see a reason for that.

dhiller replied:
(also since I am quite deeply involved in community matters, I am willing to help out to that regard)

lyarwood replied:
This was just a suggestion to help with initial communication out of the WG to the final active SIG in the list. As @EdDev said this isn't required and once things start moving forward with the WG I'm sure formal communication to all of the SIGs will be more than adequate.

lyarwood commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,64 @@
+
+With this in mind, there is a need to define a clear scope on which areas and through what means this can be accomplished.
+
+* Create and maintain code-quality guidelines, best practices and standards.
```
nit - link to https://github.com/kubevirt/kubevirt/blob/main/docs/coding-conventions.md ?

EdDev replied:
The link points to *how" which I think should be left out of a charter.
Here we just want to say that creating and maintaining such a convention is in the WG scope, that is it.

lyarwood commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,64 @@
+
+The Code-Quality Working Group has two goals:
+
+* (short-term) **focus on the improvement of code quality** and
```
nit - drop the `and` at the end

lyarwood reviewed (commented):
Quick comments, overall this looks great and apologies for the lack of review here.

EdDev reviewed (approved):
@dhiller , thank you very much for pushing this forward.

dhiller commented:
I did a final polish, hope that's good enough for now - @orelmisan @lyarwood @EdDev @iholder101 PTAL again.

Thanks all for your input!

@aburdenthehand this should be nearing the approval phase.

orelmisan reviewed (approved):
Thank you @dhiller!

iholder101 commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,65 @@
+
+### Special Interest Groups
+
+* All SIGS claiming code ownership and producing code, currently
+  * SIG Compute
+  * SIG Network
+  * SIG Storage
```
nit: observability? testing? documentation?
which other SIGs do we currently have? :D

aburdenthehand replied:
This makes sense for now, based on SIGs with current charters, although SIG-observability has an [open PR](https://github.com/kubevirt/community/pull/353) so it would be worth including them.
Is the intention to update this as other SIGs are formalised?

iholder101 reviewed (commented):
Thanks a lot @dhiller!!

/lgtm

EdDev commented:
> /hold to wait for rebase after https://github.com/kubevirt/community/pull/346 merged

It has been merged already.
/unhold

@aburdenthehand , can we merge this?

lyarwood reviewed (approved):
/lgtm

aburdenthehand commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,65 @@
+  * SIG Compute
+  * SIG Network
+  * SIG Storage
+* sig-ci
```
```suggestion
* SIG CI
```

aburdenthehand commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,65 @@
+* Create and maintain code-quality guidelines, best practices and standards.
+* Assure distribution of knowledge in regard to code quality.
+  * This may include presentations, documentation, lectures, certifications, linters, etc.
+* Enforcement of selected standards.
+  * Through automation and active intervention in code reviews.
+  * Through codebase infrastructure. E.g reusable libraries.
```
```suggestion
* Enforcement of selected standards via the following methods:
  * Automation and active intervention in code reviews.
  * Codebase infrastructure. E.g reusable libraries.
```

aburdenthehand commented on wg-code-quality/wg-charter.md:
```diff
@@ -0,0 +1,65 @@
+
+### Special Interest Groups
+
+* All SIGS claiming code ownership and producing code, currently
```
```suggestion
* All SIGs claiming code ownership and producing code, currently
```

aburdenthehand reviewed (commented):
Good stuff. I have a few suggestions and an additional comment to Itamar's.

aburdenthehand commented:
/approve
